src/network/bufferedconnection.h

```diff
@@ -0,0 +1,321 @@
+	 */
+	bool is_connected() const;
+
+	/**
+	 * Closes the connection.
+	 * If you want to send a goodbye-message, do so before calling this.
+	 */
+	void close();
+
+	/**
+	 * Receive a command.
+	 * \warning Calling this method is only safe when peek_cmd() returned \c true.
+	 *          Otherwise the behavior of this method is undefined.
+	 * \param out The variable to write the value to.
+	 */
+	void receive(RelayCommand* out);
+
+	/**
+	 * Receive an uint8_t.
+	 * \warning Calling this method is only safe when peek_uint8_t() returned \c true.
+	 *          Otherwise the behavior of this method is undefined.
+	 * \param out The variable to write the value to.
+	 */
+	void receive(uint8_t* out);
+
+	/**
+	 * Receive a string.
+	 * \warning Calling this method is only safe when peek_string() returned \c true.
+	 *          Otherwise the behavior of this method is undefined.
+	 * \param out The variable to write the value to.
+	 */
+	void receive(std::string* out);
+
+	/**
+	 * Receive a RecvPacket.
+	 * \warning Calling this method is only safe when peek_recvpacket() returned \c true.
+	 *          Otherwise the behavior of this method is undefined.
+	 * \param out The variable to write the value to.
+	 */
+	void receive(RecvPacket* out);
+
+	// Temporary method, will be removed.
+	// Removes a message from type kRoundTripTimeResponse from the buffer.
+	void ignore_rtt_response();
+
+	/**
+	 * Sends data over the connection.
+	 * @note All data that belongs together has to be send with one function call. If added with
+	 *       multiple function calls the packet might be disrupted by a higher priority packet,
+	 *       breaking the network protocol and leading to a disconnect.
+	 * @param priority The priority this data has. A higher priority can be transmitted earlier,
+	 *                 even when a large group of low priority data has been scheduled earlier.
+	 * @param Fargs A list of data objects that should be send as part of this packet.
+	 *              The following data types can be send:
+	 *              RelayCommand, uint8_t, std::string, std::vector<uint8_t>, SendPacket.
+	 */
+	/*
+	 This method creates a vector to store the data-to-be-send in. Basically this template magic
+	 (called "parameter pack") is similar to the var-arg magic used by printf(). An arbitrary
+	 number of arbitrary-typed parameters might be passed as the Fargs parameter. The compiler than
+	 decides which of the send_T_() methods have to be called. Since each of them takes a different
+	 first argument out of Fargs only one of them can match. The matching methods transforms the
+	 received first argument to an uint8_t string and appends it to the given vector. Then it
+	 calls send_T_() again, but with one argument less. This results in a "recursive" call
+	 until no more arguments are left.
+	 */
+	template<typename... Targs>
+	void send(NetPriority priority, const Targs&... Fargs) {
+
+		std::vector<uint8_t> v;
```

Comment:
TODO(k.khalfmann): think about some reasobale presizing?

Reply:
Sounds sensible, done.

Reply:
std::vector does not allow  presizing in the CTor? like `std::vector<uint8_t> v(kNetworkBufferSize);`
Indeed, so thats OK now.

Reply:
@klaus-halfmann Checking out from a fork is a lot easier than that - see https://www.widelands.org/wiki/GitPrimer/#checking-out-a-branch-from-somebody-elses-fork